### ISSUE TITLE
Requesting Nitro-E

### ISSUE THREAD

**Issue by jdluzen** [TensorStack-AI/TensorStack#1]
Great to have you back Adam!
I'd like to request [Nitro-E](https://github.com/AMD-AGI/Nitro-E) support. I see some pieces are already supported, like `FlowMatchEulerDiscreteScheduler`, and things related to `MMDiT`. Is that reusable from SD3? I just started looking at the Nitro-E repo and comparing it to TensorStack. So much bigger since the last one.

Models:
`meta-llama/Llama-3.2-1B`, ONNX: https://huggingface.co/onnx-community/Llama-3.2-1B
`mit-han-lab/dc-ae-f32c32-sana-1.0-diffusers`, ONNX: available? I think, maybe. After much struggling, I managed to get https://github.com/mit-han-lab/efficientvit/blob/master/assets/onnx_export.py working. But I'm not sure what it exported, it's only 243MB, when the original is 1.25GB. I did not intentionally quantize it.

**Comment by saddam213** [comment]
Hey Man :)

Looks like there is some onnx versions of Llama-3.2-1B in that link, however we will need a TextGeneration pipeline as this is a decoder model not a basic TextEncoder

I have created a repo with the onnx model and a tokenizer format that is compatible with TensorStack
[TensorStack/Llama-3.2-1B](https://huggingface.co/TensorStack/Llama-3.2-1B)

A Llama pipeline should be fairly trivial as TensorStack already supports decoder only models


Converting the diffusion transformer model and vae encoder is an unknown, without that adding support is not likely, I have no clue how to convert models like this

**Comment by saddam213** [comment]
[Llama-3.2-1B](https://huggingface.co/TensorStack/Llama-3.2-1B) support added: https://github.com/TensorStack-AI/TensorStack/commit/7559adf2effe5aed38656d857e47405e761e125f

Hopefully someone clever can convert the rest to onnx

**Comment by jdluzen** [comment]
> I cant run the python example as some of the models are gated
> 
> ```
> Cannot access gated repo for url https://huggingface.co/meta-llama/Llama-3.2-1B/resolve/main/config.json.
> Access to model meta-llama/Llama-3.2-1B is restricted. You must have access to it and be authenticated to access it. Please log in.
> ```
> 
> Not keen to give meta my data for this

I know the feeling. `unsloth/Llama-3.2-1B` works out of the box for me.

**Comment by jdluzen** [comment]
I also only have AMD. Though I am using WSL with a custom Dockerfile and ROCm, Nitro is so fast that you can change your device from `cuda` to `cpu` and barely notice.

**Comment by jdluzen** [comment]
Wow great! What's your `torch.onnx.export` look like? Mine's broken somehow. I'm currently stuck on the tensor shapes for the transformer. I'm currently using `(torch.randn(768, 32, 1, 1, dtype=dtype), torch.randn(1, 128, 2048, dtype=dtype), torch.tensor([999], dtype=torch.long))` which isn't right.

**Comment by saddam213** [comment]
im just editing one of the amuse converters
https://github.com/TensorStack-AI/AmuseConverters/tree/main/StableDiffusion3

will upload a Nitro one soon, well what I have so far

**Comment by saddam213** [comment]
latent_model_input = [1, 32, 16, 16]
encoder_hidden_states = [1, 128, 2048]
encoder_attention_mask = [1, 128]
timestep = [1]

**Comment by jdluzen** [comment]
Those certainly looks a lot better than mine 😅

**Comment by saddam213** [comment]
ok, think I have it sorted, have converted the 512px model

<img width="3368" height="1491" alt="Image" src="https://github.com/user-attachments/assets/883529fd-fe04-4a82-83f5-fa4f325818d8" />


Now have to make a pipeline to put it all together and see if it work

I am stuck in bed injured ATM, but hopefully I can get this done over the next few days

**Comment by saddam213** [comment]
I have not had much luck with the AutoEncoder either, I have the transformer working on CPU (float32), well its returning data anyway
https://huggingface.co/TensorStack/Nitro-E

The c# pipeline is compete, it will work once we have the models as the shapes are clear now, so I will merge the PR but keep this issue open and hopefully someone can figure out the onnx models

**Comment by saddam213** [comment]
It most likely wont be possible unfortunately, onnx is just too limited for stuff like this, none of the video model can be converted either specifically the VAE decoders

**Comment by saddam213** [comment]
I followed the recommendation from the Nitro team and modified the 3 models they mentions to replace `movedim` with `permute`
https://github.com/AMD-AGI/Nitro-E/issues/6

The `vae_decoder` no longer returns `NAN` or an error WOOT!!, 

However the output is incorrect.

"Cute Cat"
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/72e93482-b726-4576-9186-abe2e8966882" />

Not sure whats causing the issue yet, `vae_decoder`, `transformer` or `text_encoder`

And still not able to convert to flaot16, but will worry about that once we get float32 working

**Comment by saddam213** [comment]
OK, seems I picked the wrong node for `last_hidden_state` in the `text_encoder`

"Cute Cat" (512px, 4 steps, float32)
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/cf78c176-1efd-44c4-8593-f4a6187d8ca8" />
(I expect quality to improve as the padding is incorrect right now)

I have updated huggingface with the 512 4 step model
https://huggingface.co/TensorStack/Nitro-E

I will work on the 1024 models now and then see if I can get them all converted down to float16

**Comment by saddam213** [comment]
you just need to edit the python model files

Onnx export takes the weights from the safetensor files and then parses the diffusers python model and tries to replicate it with onnx ops

a lot of times you need to edit the diffusers python models to make things onnx friendly, onnx only supports tensors as inputs, so a lot of times you need to update inputs from scalar or list to a tensor to get a conversion to work

but also onnx is missing a LOT of operations that exist in diffusers models, so this makes exports impossible in some cases


so here I just edited the model file, then re-ran the export
<img width="2862" height="1881" alt="Image" src="https://github.com/user-attachments/assets/b06bb8ce-bd0d-4106-ad76-f5bce8261830" />
